### PR TITLE
Introduce SignedArtifactBundle

### PR DESCRIPTION
**Pull request by danbev**
#### Summary

This commit adds a new struct named SignedArtifactBundle which represents a signed artifact bundle as produced when signing with cosign and specifying the --bundle option.

The motivation for change comes from trying to implement an example that verifies a blob using a bundle.
For example, first a blob is signed using the following command:
```console
cosign sign-blob --bundle=artifact.bundle artifact.txt
```
The `artifact.bundle` file generated by the above command will look something like this (shortened to fit the commit message format):
```json
{
  "base64Signature": "...",
  "cert": "...",
  "rekorBundle": {
    "SignedEntryTimestamp": "...",
    "Payload": {
      "body": "...",
      "integratedTime": 1669361833,
      "logIndex": 7810348,
      "logID": "..."
    }
  }
}
```

Currently, to create Bundle from this, one would have to parse the string as json, and then access the `rekorBundle` element, and then serialize it so that it can be passed to `Bundle::new_verified`.

With the changes in this commit it will be possible to call `SignedArtifactBundle::new_verified` and pass in the contents for the bundle file directly.

Refs: https://github.com/sigstore/sigstore-rs/issues/117
Signed-off-by: Daniel Bevenius <daniel.bevenius@gmail.com>

#### Release Note

NONE

#### Documentation

NONE

### CODE REVIEW

File: src/cosign/bundle.rs [hunk 1]
```diff
@@ -20,23 +20,52 @@ use std::cmp::PartialEq;
 use crate::crypto::{CosignVerificationKey, Signature};
 use crate::errors::{Result, SigstoreError};
 
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Bundle {
+    #[serde(rename(deserialize = "base64Signature"))]
+    pub base64_signature: String,
+    pub cert: String,
+    #[serde(rename(deserialize = "rekorBundle"))]
```
Comment (viccuad) [inline src/cosign/bundle.rs:28]: maybe both deserialize can be generalized for the whole struct with `#[serde(rename_all = "camelCase")]`.

**Review by viccuad — APPROVED** [review]
Thanks for the contribution! looks good to me, modulo the serde nitpick.

File: src/cosign/bundle.rs (hunk 1 above)
Comment (danbev) [reply src/cosign/bundle.rs:28]: Ah good point! I'll update this shortly. Thanks!

**Review by viccuad — DISMISSED** [review]

**Comment by viccuad** [comment]
Indeed, this is a breaking change for `SignatureLayer{}` if there's a name change. I am also happy with `SignedArtifactBundle`.

**Comment by danbev** [comment]
> I have nothing against adding this new struct, but I'm a bit reluctant to rename the current `Bundle` struct. This would break our public API, which is not a taboo, but should be pondered.
> 
> Why don't we create instead a new struct called `SignedArtifactBundle`? The bundle you're referencing is created only by the `sign-artifact` command of cosign. I think this approach would be more clear about what this new struct is about and avoid the API break.
> 
> What do you think?

Sounds good, I'll take a look at this and also add some documentation :+1:

**Comment by flavio** [comment]
Thanks @danbev! :pray:

File: src/cosign/bundle.rs [hunk 2]
```diff
@@ -20,6 +20,28 @@ use std::cmp::PartialEq;
 use crate::crypto::{CosignVerificationKey, Signature};
 use crate::errors::{Result, SigstoreError};
 
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct SignedArtifactBundle {
+    pub base64_signature: String,
+    pub cert: String,
```
Comment (flavio) [inline src/cosign/bundle.rs:27]:
```suggestion
    /// PEM encoded certificate
    pub cert: String,
```

File: src/cosign/bundle.rs [hunk 3]
```diff
@@ -20,6 +20,28 @@ use std::cmp::PartialEq;
 use crate::crypto::{CosignVerificationKey, Signature};
 use crate::errors::{Result, SigstoreError};
 
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
```
[the remark this answers is not in the document]
Comment (danbev) [reply src/cosign/bundle.rs:29]: I've added some comments now. Thanks

**Review by flavio — APPROVED** [review]

**Pull request merged**